drivers/platform/linux/linux_uart.c

```diff
@@ -0,0 +1,298 @@
+/******************************************************************************/
+#include "error.h"
+#include "uart.h"
+#include "linux_uart.h"
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct linux_uart_desc
+ * @brief Linux platform specific UART descriptor
+ */
+struct linux_uart_desc {
+	/** /dev/"device_id" file descriptor */
+	int fd;
+	struct termios *terminal;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Initialize the UART communication peripheral.
+ * @param desc - The UART descriptor.
+ * @param param - The structure that contains the UART parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t uart_init(struct uart_desc **desc, struct uart_init_param *param)
+{
+	struct linux_uart_init_param *linux_init;
+	struct linux_uart_desc *linux_desc;
+	struct uart_desc *descriptor;
+	speed_t speed;
+	char path[64];
+	int ret;
+
+	descriptor = malloc(sizeof(*descriptor));
+	if (!descriptor)
+		return FAILURE;
+
+	linux_desc = (struct linux_uart_desc*) malloc(sizeof(struct linux_uart_desc));
+	if (!linux_desc)
+		goto free_desc;
+
+	linux_desc->terminal = (struct termios*) malloc(sizeof(struct termios));
+	if (!linux_desc->terminal)
+		goto free_linux_desc;
+
+	descriptor->extra = linux_desc;
+	linux_init = param->extra;
+
+	ret = snprintf(path, sizeof(path), "/dev/%s", linux_init->device_id);
+	if (ret < 0 && ret >= sizeof(path))
+		goto free_linux_desc;
+
+	linux_desc->fd = open(path, O_RDWR | O_NOCTTY | O_NONBLOCK);
+	if (linux_desc->fd < 0) {
+		printf("%s: Can't open %s\n\r", __func__, path);
+		goto free_terminal;
+	}
+
+	tcgetattr(linux_desc->fd, linux_desc->terminal);
+
+	cfmakeraw(linux_desc->terminal);
+
+	switch(param->baud_rate) {
+		case 50:
+			speed = B50;
+			break;
+		case 75:
+			speed = B75;
+			break;
+		case 110:
+			speed = B110;
+			break;
+		case 134:
+			speed = B134;
+			break;
+		case 150:
+			speed = B150;
+			break;
+		case 200:
+			speed = B200;
+			break;
+		case 300:
+			speed = B300;
+			break;
+		case 600:
+			speed = B600;
+			break;
+		case 1200:
+			speed = B1200;
+			break;
+		case 1800:
+			speed = B1800;
+			break;
+		case 2400:
+			speed = B2400;
+			break;
+		case 4800:
+			speed = B4800;
+			break;
+		case 9600:
+			speed = B9600;
+			break;
+		case 19200:
+			speed = B19200;
+			break;
+		case 38400:
+			speed = B38400;
+			break;
+		default:
+			goto free;
```

Comment:
maybe this should be:
```
ret = -EINVAL;
goto free;
```

And free: shoud end with:
```
return ret;
```

For more meaningful error handling.

Also memory allocations above could set ret = -ENOMEM etc.